### PR TITLE
Separate query type transformations

### PR DESCRIPTION
**Pull request by idastambuk**
**What this PR does / why we need it**:
1. Fixes use case when a panel has queries of different types, returns one observable for targets of each type
2. Since there can also be multiple queries of type 'traces' or 'traceId', also adds support for returning dataFrames for all of them in formatTraces.ts

If there are multiple queries in a single panel, the table will have a dropdown with all the queries: 
Screenshot 2023-05-03 at 9 49 57 AM

This is the way other plugins work, more specifically I tried X-ray for traces and Cloudwatch. 

**Details about the implementation:**
Currently  in `main`, we use multi_search to combine all requests to OpenSearch (of every type). Before we added traces support, all the query types in OpenSearch were transformed into data frames according to a certain logic. However, after adding traces, there was yet another way to transform the data, which led to problems: After the array with all the results for different queries is received, condition 
`if (targets.some(target => target.isLogsQuery)) {` prepares data frames according to the logic for logs queries, even if one of the queries was for trace list for example. This leads to the bug in the original bug report, where there can be a for example, logs query and a traces query. 

There were two approaches I could think of to fix this: 
1. Separate the responses we were getting from multi_search only when we get the response. So, pick out the results for trace list, and metrics, and logs, separate them out of the response.data array, then transform them according to their different logic. However, [this proved to be a bit difficult ](https://github.com/grafana/opensearch-datasource/pull/152)since, even though OpenSearch API returns the results IN ORDER we send the request in, it returns them in two different places: hits[] and aggregations[] fields. Since we use both fields  for some query types (trace list and metrics afai can see) to create a data frame, I estimated it would be a bit unreadable and maybe even impossible to separate them in a nice way. 
2. Separate the query types into three different observables, that will emit values with data frames transformed according to their respective query types. This way we don't have to worry about combining all the different data frame collections into one `DataQueryResponse.data` array. Since this creates 3 requests, see below. 

In this implementation observable for each query type (traces, trace, metrics + logs) emits data transformed into dataframes, as it receives it, so 

Known issues/things I dont love for with this solution:
1. In single trace table, there is a Field1, not sure why. Will take a look at it in the meantime. 
2. The request function (at its most complicated) makes 3 requests to OpenSearch. However, I haven't found proof that this makes it more expensive for the end user, but I haven't found proof otherwise either. 
However, I don't expect the multiple requests to happen very often, since I predict the users will either have one query per panel, or multiple of the same queries (so, not combining Metrics and Traces, for example)
3. There are 3 types of requests: `traceList`, `trace` and `other`. The "other" covers logs and metrics queries both. This is because I didn't want to mess with the old solution where these two (when it was the only ones that existed) were together, so I kept them. I didn't think refactoring or figuring out if there are problems with that approach is in the scope of this.
4. There are no tests for this as of yet. If we decide to merge this I will write them

**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/opensearch-datasource/issues/135

**Special notes for your reviewer**:
Suggestions for other approaches are welcome, I might have missed something that would be easier
I think this would be an interim solution until we refactor the repo and move it to the BE, so Im ok with keeping this unmerged until we start getting bug reports (since I predict users won't be having multiple queries a panel often)

### CODE REVIEW

File: src/traces/formatTraces.ts [hunk 1]
```diff
@@ -46,88 +46,99 @@ export const createListTracesDataFrame = (
...
+    series = addPreferredVisualisationType(series, 'trace');
+    dataFrames.push(series);
+    return dataFrames;
   }
-  // do we need this?
-  series.refId = targets[0].refId;
-  series = addPreferredVisualisationType(series, 'trace');
-  dataFrames.push(series);
+  // if multiple targets of type: trace, flatMap them into data(should be a single array with all trace data frames)
+  const data = targets.flatMap((target, index) => getDataFrameForTarget(target, response.responses[index].hits.hits));
```
Comment (idastambuk) [inline src/traces/formatTraces.ts:139]: since queries come in the same order that they're [sent,](https://opensearch.org/docs/latest/api-reference/multi-search/#response) I think we can count on the index of target to be the same as index of response for that target

File: src/traces/formatTraces.ts [hunk 2]
```diff
@@ -46,88 +46,99 @@ export const createListTracesDataFrame = (
...
+        { name: 'Error Count', type: FieldType.number, values: errors },
+        { name: 'Last Updated', type: FieldType.time, values: lastUpdated },
+      ],
+    };
+    return new MutableDataFrame(traceFields);
+  }
+  // if multiple targets of type traceList, map them into data
+  const dataFrames = response.map(res => createDataFrame(res));
+
+  return { data: dataFrames, key: targets[0].refId };
```
[the remark this answers is not in the document]
Comment (idastambuk) [reply src/traces/formatTraces.ts:102]: [The docs](https://github.com/grafana/grafana/blob/4ded937c79a38087576b3224bf6f2ba59345571e/packages/grafana-data/src/types/datasource.ts#L446) for "key" refer to partial responses, which I don't think we need to account for right now (and they haven't been accounted for before afaik)
HOWEVER, good catch - I wasnt passing refId to trace list frames in createListTracesDataFrame! Just committed the fix for that :) Not all DataQueryResponseData objects will have a refId

[the remark this answers is not in the document]
Comment (sarahzinger) [reply src/traces/formatTraces.ts:108]: Good find with those docs! Hmm if we're not using partial responses maybe we don't need to pass a key at all? It seems optional?

**Review by sarahzinger — COMMENTED** [review]
Hey hey! We talked about this on a video call, but IMO:
- this is a great approach and if we want to write some tests and merge it I'm game!
- i don't know that users are hitting this edge case and I think we're going to rewrite a lot of this codebase in the next quarter and if we want to tackle it then that is also a very reasonable choice! Would recommend double checking this pr is linked to the issue and then putting it in the backlog.

**Review by iwysiu — COMMENTED** [review]
(I realize that it doesn't sound like this is going to be merged anytime soon, but I was curious about the typescript syntax)
In the context of future refactors it may be good to check if this issue exists on the backend? Though from my cursory search it doesn't seem like its handling traces at all (which makes sense in the context of the backend being for alerting/expressions). If we expect that we're going to refactor opensearch to run through the backend, it may just be best to handle it when we reimplement traces in the backend.
But ignoring possible refactors I like this approach and it makes a lot of sense to me!

File: src/datasource.ts [hunk 3]
```diff
@@ -506,45 +506,74 @@ export class OpenSearchDatasource extends DataSourceApi<OpenSearchQuery, OpenSea
...
+    const traceListTargets = targets.filter(
+      target => target.luceneQueryType === LuceneQueryType.Traces && !getTraceIdFromLuceneQueryString(target.query)
+    );
+    const traceTargets = targets.filter(
+      target => target.luceneQueryType === LuceneQueryType.Traces && getTraceIdFromLuceneQueryString(target.query)
     );
+
+    const otherTargets = targets.filter(target => target.luceneQueryType !== LuceneQueryType.Traces);
+
+    const traceListObs$ =
```
[the remark this answers is not in the document]
Comment (idastambuk) [reply src/datasource.ts:540]: Good question, usually "$"s are added to distinguish regular variables from observables. But looking at these again, I renamed them from e.g. traceListObs$ to traceList$, since the names contained a redundancy ("obs" + "$")

File: src/utils.ts [hunk 4]
```diff
@@ -73,16 +73,16 @@ export async function sha256(string) {
...
   logMessageField?: string,
-  logLevelField?: string
+  logLevelField?: string,
+  timeField?: string
 ): MutableDataFrame => {
   const series = new MutableDataFrame({ fields: [] });
 
   //PPL table response should add time field only when it is part of the query response
-  if (targetType === QueryType.Lucene || isLogsRequest) {
+  if (timeField && (targetType === QueryType.Lucene || isLogsRequest)) {
```
Comment (idastambuk) [inline src/utils.ts:85]: The function was adding a random "Field 1" column to traceDetails, so i did this quick fix to not add it. All older function calls have a time field, so this shouldn't change anything. I thought about removing the rest of the conditions, but this function is being called from places in the app that (I think) still need these conditions.

File: src/datasource.test.ts [hunk 5]
```diff
@@ -1232,6 +1234,176 @@ describe('OpenSearchDatasource', function(this: any) {
       );
     });
   });
+  describe('#executeLuceneQueries', () => {
```
Comment (idastambuk) [inline src/datasource.test.ts:1237]: Tried to test this with [marble diagrams and rxjs TestRunner](https://rxjs-dev.firebaseapp.com/guide/testing/marble-testing), but couldn't get the runtime observables to match the diagrams :( I also had to turn the observables to promises since as far as I could find, that's the easiest way to test the results of the calls in jest. This is why I only ever put two query types (that result in only two observables that I could turn into promise with firstValueFrom and lastValueFrom) in the tests.

**Comment by idastambuk** [comment]
> (I realize that it doesn't sound like this is going to be merged anytime soon, but I was curious about the typescript syntax) In the context of future refactors it may be good to check if this issue exists on the backend? Though from my cursory search it doesn't seem like its handling traces at all (which makes sense in the context of the backend being for alerting/expressions). If we expect that we're going to refactor opensearch to run through the backend, it may just be best to handle it when we reimplement traces in the backend. But ignoring possible refactors I like this approach and it makes a lot of sense to me!

Good point, confirming that we decided to not support traces for Alerts/Expressions, so I dont think we have to worry about this problem there before the refactor.
As for the second point, I think we can merge this now, as it isn't a big impact fix, and definitely think about it once we move stuff to BE. I suggested we can wait to merge this until we see if users are getting the errors, but @sarahzinger made a good point that by then, the rebasing will be a nightmare and it doesn't make sense to wait with a bugfix if we have it.

**Review by sarahzinger — APPROVED** [review]

File: src/OpenSearchResponse.ts [hunk 6]
```diff
@@ -462,11 +462,11 @@ export class OpenSearchResponse {
         if (docs.length > 0) {
           let series = createEmptyDataFrame(
             propNames,
-            this.targets[0].timeField!,
             isLogsRequest,
             this.targetType,
             logMessageField,
-            logLevelField
+            logLevelField,
+            this.targets[0].timeField!
```
[the remark this answers is not in the document]
Comment (idastambuk) [reply src/OpenSearchResponse.ts:469]: It shouldn't, since I changed the order in the createEmptyDataFrame function. I tested it and the queries work the same!
